src/main/scala/ch/epfl/alcmp/net/Serde.scala

```diff
@@ -22,40 +22,78 @@ object Serde {
     override def serialize(obj: String): String = Base64.getEncoder.encodeToString(obj.getBytes(StandardCharsets.UTF_8))
     override def deserialize(data: String): String = new String(Base64.getDecoder.decode(data), StandardCharsets.UTF_8)
 
-  given Serde[RegisterMessage] with
-    override def serialize(obj: RegisterMessage): String = ???
-    override def deserialize(data: String): RegisterMessage = ???
 
+  given Serde[RegisterMessage] with
+    override def serialize(obj: RegisterMessage): String = Serdes.serialize[Int](obj.id)
+    override def deserialize(data: String): RegisterMessage = RegisterMessage(Serdes.deserialize[Int](data))
 
   given Serde[DoneMessage] with
-    override def serialize(obj: DoneMessage): String = ???
-    override def deserialize(data: String): DoneMessage = ???
+    override def serialize(obj: DoneMessage): String = Serdes.serialize[Int](obj.id)
+    override def deserialize(data: String): DoneMessage = DoneMessage(Serdes.deserialize[Int](data))
 
   given (TypeId => Serde[DivideMessage]) with
     override def apply(t: TypeId): Serde[DivideMessage] = new Serde[DivideMessage]:
-      override def serialize(obj: DivideMessage): String = t match
-        case TypeId.ListType => ???
-        case TypeId.MatrixType => ???
-        case TypeId.HeapType => ???
-        case TypeId.BinaryTreeType => ???
-      override def deserialize(data: String): DivideMessage = t match
-        case TypeId.ListType => ???
-        case TypeId.MatrixType => ???
-        case TypeId.HeapType => ???
+      override def serialize(obj: DivideMessage): String =
+        val s = List(obj.id, obj.depth, obj.index, obj.highlights.mkString(",")).mkString("/") + "/"
+        val b1 = StringJoiner(";")
+        for ( output <- obj.outputs) {
+          output match
+            case IList(list) => b1.add(list.mkString(","))
+            case IMatrix(rows) =>
+              val b2 = StringJoiner(",,")
+              rows.foreach(list => b2.add(list.mkString(",")))
+              b1.add(b2.toString)
+            case IHeap(list) => b1.add(list.mkString(","))
+            case IBinaryTree(root) => ???
+        }
+        s + b1.toString
+
+      override def deserialize(data: String): DivideMessage =
+        val arr = data.split("/")
+        val (id, depth, index) = (arr(0).toInt, arr(1).toInt, arr(2).toInt)
+        val highlights = arr(3).split(",").map(s => s.toInt).toList
+        t match
+        case TypeId.ListType =>
+          val outputs = arr(4).split(";")
+                        .map(list => IList(list.split(",").map(s2 => s2.toInt).toList)).toList
```

Comment:
Same here, I think some composition with the Serde I defined above may make the code easier to understand and less of a hassle to update if we decide to change the protocol